### PR TITLE
Reflection_Engine: extraction of CRUD methods compatible with specific objects types

### LINKED ISSUE #1487: Reflection_Engine: implement extraction of CRUD methods compatible with specific objects types

### PR DESCRIPTION
**Pull request by alelom**
### Issues addressed by this PR

Closes #1487 

Implemented methods as described by #1487.

The main method we are looking for here is [`CRUDcompatibleTypes()`](https://github.com/BHoM/BHoM_Engine/pull/1489/files#r382097411). This returns a `Dictionary<string,  Dictionary<Type, List<MethodInfo>>>`, that has:
- string key, one of: `Create`, `Read`, `Update`, `Delete`;
- value: another Dictionary with:
   - key: BHoM Type of with some CRUD method that supports it
   - value: list of CRUD methods compatible with that Type.

### Test files

@FraserGreenroyd to test

### CODE REVIEW

File: Reflection_Engine/Query/AdapterMethods.cs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+            return GroupedCRUDmethods;
+        }
+
+        /***************************************************/
+
+        [Description("Returns a collection of all Types that are supported by some CRUD action, grouped per object type.")]
+        [Output("Dictionary with string key, one of: `Create`, `Read`, `Update`, `Delete`;\n" +
+            "value is a Dictionary with key: BHoM Type of with some CRUD method that supports it; value: list of CRUD methods compatible with that Type.")]
+        public static Dictionary<string, Dictionary<Type, List<MethodInfo>>> CRUDcompatibleTypes()
```
Comment (alelom) [inline Reflection_Engine/Query/AdapterMethods.cs]: Main method of interests here.

**Comment by FraserGreenroyd** [comment]
This is looking good @alelom from a brief code review I've done. I'll do a full proper test with the API stuff as well to see if that works for that use case ASAP :smile:

**Comment by FraserGreenroyd** [comment]
> * I am wondering why we are not formalising the name of `Create` methods. We would end up with a single string for each category instead of a list. 3 of them should probably not be handled as list as you don't pass them to another generic method and therefore have no need to have a single format.

Maybe something for discussion in the Framework meeting?

> * I found slightly odd that this code is in the Reflection_Engine as this is purely about adapters.

I think we opted to put it here (at least initially) because it was meant to complement the BHoMMethodList() type functionality to reflect out the specific adapter methods in addition to engine methods/objects.

> So, overall, I am ok if this is merged as is but I would prefer if it was either moved to the adapter engine or made more generic to better suit the purpose of the reflection engine.

I think it would be worth keeping this in Reflection_Engine to keep it together with the rest of the reflection code, so I'd be happy for the API to wait a little longer to get this done in a more generic way to suit this engine than move it to Adapter_Engine. But I also have no real qualms with working with it from there if we do move it - so I guess I'm happy either way ultimately but with a slight bias towards here :smile:

File: Reflection_Engine/Query/AdapterMethods.cs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+using BH.oM.Base;
+using BH.oM.Reflection.Attributes;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Reflection;
+
+
```
Comment (FraserGreenroyd) [inline Reflection_Engine/Query/AdapterMethods.cs]: Remove blank line

File: Reflection_Engine/Query/AdapterMethods.cs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                        result[firstPar.ParameterType] = new List<MethodInfo>() { createMethods[i] };
+                    else
+                        result[firstPar.ParameterType].Add(createMethods[i]);
+                }
+            }
+
+            return result;
+        }
+
+
```
Comment (FraserGreenroyd) [inline Reflection_Engine/Query/AdapterMethods.cs]: Remove blank line

File: Reflection_Engine/Query/AdapterMethods.cs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+                        result[returnType] = new List<MethodInfo>() { readMethods[i] };
+                    else
+                        result[returnType].Add(readMethods[i]);
+                }
+            }
+
+            return result;
+        }
+
+
```
Comment (FraserGreenroyd) [inline Reflection_Engine/Query/AdapterMethods.cs]: Remove blank line

**Comment by adecler** [comment]
I've only had a quick look but I'll be honest. I like this version even less than the last version I reviewed. I don't have the time to properly review as there are too many urgent things that have popped up so I won't block it if you absolutely need to merge but we'll need to come back to this as a minimum.

**Comment by al-fisher** [comment]
I have had a quick look at this - and I think needs more input from you @adecler 
Can I give a bump to this as needs to be closed this sprint as part of MVP? - is critical for API docs. Ta!

@adecler @alelom can you grab a short chat to close out before early/mid next week?
@adecler you also mention happy to merge to unblock and revisit. I am relaxed either way. 
Thanks both
@FraserGreenroyd FYI RE API, LOL

**Comment by adecler** [comment]
Invite sent to Alessio for a chat on Tuesday.

**Comment by al-fisher** [comment]
Awesome! Thanks guys! 😄

**Comment by alelom** [comment]
This is now replaced by https://github.com/BHoM/BHoM_Adapter/pull/212.

**Pull request closed without merging**